Add Dockerfile and instructions

camerondurham opened the pull request:
Hi there!

I've added a Dockerfile that makes a *relatively small* multi-stage image of the CLI that should be usable cross-platform (only tested on macOS so far but no reason to believe it would fail on Windows/Linux).

I'm interested in contributing something more meaningful (i.e. tests for #10 when I have a little more time) but wanted to send this for now. I also understand if you don't want to merge since it *does* sadly change your repo stats from  100% to 99.1% Rust, 0.9% Dockerfile :smile:.

Question for maintainers:
- This builds and appears to run with latest release of Rust as of 10/22, version 1.56. Should this be pinned to another Rust version?
- I haven't looked too closely into [amzn/ion-c](https://github.com/amzn/ion-c) CMake files but the only C/C++ build dependencies on Debian appeared to be `cmake` and `clang`. Should I update the README to reflect this or Dockerfile be sufficient?

Thanks!


*Issue #, if available:*

N/A

*Description of changes:*

* Added Dockerfile to build minimal, multi-stage image
* Updated README with build instructions

Recording of local build/test:

[![asciicast](https://asciinema.org/a/Ql7Xc5fgxxvkFrjjcuMbR0npY.svg)](https://asciinema.org/a/Ql7Xc5fgxxvkFrjjcuMbR0npY)

Currently builds image size 85.8MB from local build on macOS

```
docker images
REPOSITORY                               TAG               IMAGE ID       CREATED          SIZE
ion-cli                                  0.1.1             e81a9d863b95   49 minutes ago   85.8MB
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

zslayton commented:
Thanks, @camerondurham! This looks great. Getting `ion-cli` to build can be a bit of a hassle right now, so having an easy Dockerized path reduces the barrier to entry for folks looking to experiment.

> This builds and appears to run with latest release of Rust as of 10/22, version 1.56. Should this be pinned to another Rust version?

v1.56 is fine, though as of today there's a [v1.56.1](https://blog.rust-lang.org/2021/11/01/Rust-1.56.1.html). Does 

> I haven't looked too closely into amzn/ion-c CMake files but the only C/C++ build dependencies on Debian appeared to be cmake and clang. Should I update the README to reflect this or Dockerfile be sufficient?

I think it's worth mentioning in the `README`.

zslayton commented on Dockerfile:
```diff
@@ -0,0 +1,13 @@
+RUN apt-get update -y \
+  && apt-get install -y ${builddeps} \
+  && git submodule update --init --recursive
+RUN cargo install --path .
```
It'd be nice to make sure this configuration works each time we commit new code. How easy would it be to add building this Docker image to the GitHub actions?

camerondurham replied:
Agreed, this should definitely have an action.

zslayton commented on README.md:
```diff
@@ -33,6 +33,39 @@ and the API is subject to change._
+   # print the help message
+   docker run -it --rm ion-cli:0.1.1 ion -V
+
+   # mount current directory to /data volume and dump an ion file
+   docker run -it --rm -v $PWD:/data ion-cli:0.1.1 ion dump /data/test.ion
```
Thanks for including this example, it's really helpful. 👍

zslayton commented on README.md:
````diff
@@ -33,6 +33,39 @@ and the API is subject to change._
+   ```
+   cd ion-cli
+   ```
+3. Install Docker (see OS specific instructions on the [Docker website](https://docs.docker.com/get-docker/))
````
I'd make this step 1. If the host system needs other libraries (Debian needs the host to have `cmake` and `clang`, you mentioned?) then you could turn this step into `Install Dependencies` with a sublist.

camerondurham replied:
Good point, this should be re-ordered. 

For the Docker specific instructions, the only host dependency will be Docker. Sorry it was confusing how I mentioned the `clang` and `cmake` deps. I wanted to call out that the host system appears to only need those.

camerondurham commented:
Thanks for the review and comments @zslayton !

Revision summary:

1. added Github workflow that will run on PRs and push to `master` branches
2. re-ordered Docker usage instructions
3. added note about build dependencies
4. updated to latest `1.56.1` Rust Docker image


On my fork, the workflow is working as expected.

Summary in repo commits [here](https://github.com/camerondurham/ion-cli/commits/master):

1. Successful workflow: https://github.com/camerondurham/ion-cli/actions/runs/1410831208
2. Failed workflow (removed `clang`): https://github.com/camerondurham/ion-cli/actions/runs/1410847829
3. Fixed build again: https://github.com/camerondurham/ion-cli/actions/runs/1410855706

I am sure you are aware of this, but this repo's action policy should maybe limit when actions run for certain contributors if it isn't already? [Docs: Configuring Required Approval For Workflows](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#configuring-required-approval-for-workflows-from-public-forks)

zslayton commented on README.md:
```diff
@@ -24,7 +24,7 @@ and the API is subject to change._
    This will put a copy of the `ion` executable in `~/.cargo/bin`.
 
-   **If this step fails:** You're likely missing one of `ion-c`'s dependencies. Make sure you have `cmake`, `gcc`, `g++`, and `libc++` installed.
+   **If this step fails:** You're likely missing one of `ion-c`'s dependencies. Make sure you have `cmake`, `gcc`, `g++` and `clang` installed. On Debian-based Linux distributions, the only required dependencies are `cmake` and `clang`.
```
Something about this isn't quite right; we shouldn't require both `gcc` _and_ `clang`. I don't want this to block the PR, though, so I'll open an issue to track it.

zslayton replied:
Opened #14.

camerondurham replied:
Agreed, that that doesn't seem right. I did only require `clang` for the Debian build but hadn't done much work to know if `gcc` was needed when building on macOS/Windows.

zslayton reviewed (approved):
This looks great, thanks for contributing!
